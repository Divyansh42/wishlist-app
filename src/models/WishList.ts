import { types, Instance, getParent, destroy, cast } from "mobx-state-tree"
const data = {
    "name" : "Headphone",
    "price" : 999,
    "image" : "https://www.amazon.in/pTron-Bassbuds-Wireless-Bluetooth-Headphones/dp/B07WSHWNH8?ref_=Oct_DLandingS_D_d157f3d0_60&smid=A14CZOWI0VEHLG"
}
export const WishListItem = types.model({
    name: types.string,
    price: types.number,
    image: ""
})
.actions(self => ({
    changeName(newName: string) {
        self.name = newName
    },
    changePrice(newPrice: number) {
        self.price = newPrice
    },
    changeImage(newImage: string) {
        self.image = newImage
    },
    remove() {
        const wishListParent = getParent<IWishList>(self, 2);
        wishListParent.remove(self);
    
    }
}))

export const WishList = types.model({
    items: types.optional(types.array(WishListItem), [])

})
.actions(self => ({
    add(item: any) {
        self.items.push(item)
    },
    remove(item: IWishListItem) {
        destroy(item);
    }
}))
.views(self => ({
    get totalPrice() {
        return self.items.reduce((sum: number, entry: IWishListItem) => sum + entry.price, 0)
    }
}))

type IWishList = Instance<typeof WishList>;

export interface IWishListItem {
    name: string;
    price: number;
    image?: string;
  } 