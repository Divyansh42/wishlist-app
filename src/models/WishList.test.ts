import { getSnapshot, onSnapshot, onPatch } from "mobx-state-tree"
import { WishList, WishListItem } from "./WishList"
import { snapshotProcessor } from "mobx-state-tree/dist/internal";

// test WishListItem
it("can create an instance of model", () => {
    const item = WishListItem.create({
        "name" : "Headphone",
        "price" : 999,
        "image" : "https://www.amazon.in/pTron-Bassbuds-Wireless-Bluetooth-Headphones/dp/B07WSHWNH8?ref_=Oct_DLandingS_D_d157f3d0_60&smid=A14CZOWI0VEHLG"
    });
    expect(item.price).toBe(999);
    item.changeName("Wireless Earphone")
    expect(item.name).toBe("Wireless Earphone")
});

// test WishList
it("can create a wishlist", () => {
    const list = WishList.create({
        items: [

            {
                "name" : "Headphone",
                "price" : 999
            }
        ]
    })

    expect(list.items.length).toBe(1)
    expect(list.items[0].price).toBe(999)
})

/**
 * using Snapshot from MST to match the whole snapshot
 */

it("can add new items", () => {
    const list = WishList.create();
    list.add({
        name: "Cheese",
        price: 100,
    });

    expect(list.items.length).toBe(1);
    expect(list.items[0].name).toBe("Cheese")
    list.items[0].changeName("Pizza Cheese")
    expect(list.items[0].name).toBe("Pizza Cheese")

    expect(getSnapshot(list)).toEqual({
        items: [
            {
                name: "Pizza Cheese",
                price: 100,
                image: "",
            },
        ],
    })

    expect(getSnapshot(list)).toMatchSnapshot();
})

/**
 * using OnSnapshot
 */

it("can add new items using OnSnapshot", () => {
    const list = WishList.create();
    const states: any = []
    onSnapshot(list, snapshot => {
        states.push(snapshot)
    });
    list.add({
        name: "Cheese",
        price: 100,
        image: "",
    });

    expect(list.items.length).toBe(1);
    expect(list.items[0].name).toBe("Cheese")
    list.items[0].changeName("Pizza Cheese")
    expect(list.items[0].name).toBe("Pizza Cheese")

    expect(getSnapshot(list)).toEqual({
        items: [
            {
                name: "Pizza Cheese",
                price: 100,
                image: "",
            },
        ],
    });
    // to MatchSnapshot creates a snapshot in the current dir
    //obtain immutable snapshots of the state using getSnapshot
    expect(getSnapshot(list)).toMatchSnapshot();
    expect(states).toMatchSnapshot()
})

it("can add new item using onPatch", () => {
    const list = WishList.create()
    const patches: any  = []
    onPatch(list, patch => {
        patches.push(patch)
    })
    list.add({
        name: "Cheese",
        price: 100, 
        image: "",
    })

    list.items[0].changeName("Pizza Cheese");

    expect(patches).toMatchSnapshot();
});