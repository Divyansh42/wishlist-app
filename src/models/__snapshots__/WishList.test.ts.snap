// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add new item using onPatch 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "image": "",
      "name": "Cheese",
      "price": 100,
    },
  },
  Object {
    "op": "replace",
    "path": "/items/0/name",
    "value": "Pizza Cheese",
  },
]
`;

exports[`can add new items 1`] = `
Object {
  "items": Array [
    Object {
      "image": "",
      "name": "Pizza Cheese",
      "price": 100,
    },
  ],
}
`;

exports[`can add new items using OnSnapshot 1`] = `
Object {
  "items": Array [
    Object {
      "image": "",
      "name": "Pizza Cheese",
      "price": 100,
    },
  ],
}
`;

exports[`can add new items using OnSnapshot 2`] = `
Array [
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "Cheese",
        "price": 100,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "Pizza Cheese",
        "price": 100,
      },
    ],
  },
]
`;
